### PR TITLE
Project Support

### PR DESCRIPTION
**Pull request by stevezhang1999**
## Update Jul. 23th

- fully support project: remove ontology from model Document
- some changes requsted are resolved.

**Important**
`fetchDocument(id)`  is deprecated, now `fetchDocument(id)` should be replaced by `fetchDocOntology(id)`.
given id of a document, `fetchDocOntology(id)` returns text of the document and corresponding ontology, whose input and return formats are identical to the corresponding formats of `fetchDocument(id)` 

**Note**
the deletion, in the changes requested, is not resolved yet.

## Original PR
This PR adds project support. 

### Description of changes

- Data models:
  - data model changed:
    - Document: add relationship with project
  - new data model 
    - Project: name, ontology and relationship with document
  -  note: ontology is not removed from Document yet
- New pages in the front end:
  - Projects - list all projects, create new project
  - Project - list all documents of the project, create documents.
  - Ontology - edit or view ontology from document (unfinished, possibly useless)
  - Signup (possibly useless)
- Page modified:
  - Document, to support new data model
- APIs in backend and frontend
  - backend: handlers added and modified
  - frontend: project-related APIs
- Miscellaneous
  - App.tsx modified
  - urls added
  - simple test case in `test.py`

### Possible influences of this PR

No side-effect provisioned.

Need more work to fully support project.

### Demonstration of results
image
image

### CODE REVIEW

File: simple-backend/nlpviewer_backend/handlers/document.py [hunk 1]
```diff
@@ -17,12 +17,18 @@ def listAll(request):
 @require_login
 def create(request):
     received_json_data = json.loads(request.body)
+    print(received_json_data)
```
Comment (hunterhector) [inline simple-backend/nlpviewer_backend/handlers/document.py:20]: Either delete these print statements, or replacing with logging.

File: simple-backend/nlpviewer_backend/handlers/document.py [hunk 2]
```diff
@@ -17,12 +17,18 @@ def listAll(request):
...
     doc = Document(
         name=received_json_data.get('name'),
         textPack=received_json_data.get('textPack'),
-        ontology=received_json_data.get('ontology')
+        ontology = received_json_data.get('ontology'),
+        project = Project.objects.get(
+            pk=received_json_data.get('project_id')
+        )
     )
+    print(received_json_data.get('project_id'))
```
Comment (hunterhector) [inline simple-backend/nlpviewer_backend/handlers/document.py:30]: same as above.

File: simple-backend/nlpviewer_backend/handlers/project.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    )
+
+    project.save()
+
+    return JsonResponse({"id": project.id}, safe=False)
+
+
+@require_login
+def edit(request, project_id):
+    pro = Project.objects.get(pk=project_id)
```
Comment (hunterhector) [inline simple-backend/nlpviewer_backend/handlers/project.py:34]: just use full name "project", no need to use abbreviations here.

File: simple-backend/nlpviewer_backend/models.py [hunk 4]
```diff
@@ -1,11 +1,44 @@
 from django.db import models
 
+# class Ontology(models.Model):
+#     ontology_name = models.CharField(max_length=200)
+#     ontology_content = models.TextField(default='')
+
+class Project(models.Model):
+    name = models.CharField(max_length=200)
+    ontology = models.TextField(default='')
+    # ontology = models.TextField(default='')
```
Comment (hunterhector) [inline simple-backend/nlpviewer_backend/models.py:10]: Remove the unused commented lines.

File: src/app/lib/api.ts [hunk 5]
```diff
@@ -19,22 +23,48 @@ export function updateDocument(id: string, name: string, textPack: string) {
...
+    name: name,
+    ontology: ontology,
   }).then(r => r.json());
 }
 
 export function deleteDocument(id: string) {
   return postData(`/api/documents/${id}/delete`);
 }
 
+export function deleteProject(id: string) {
```
Comment (hunterhector) [inline src/app/lib/api.ts:60]: what happens when the project is deleted but the documents are not?

File: src/app/pages/Ontology.tsx [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        }
+        else{
+            console.log('Save Failed'); 
+        }
+        
+       });   
+      }
+  }
+
+  console.log(ontology);
```
Comment (hunterhector) [inline src/app/pages/Ontology.tsx:58]: I think you can remove console calls after your own debug.

File: src/app/pages/Project.tsx [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+  }
+
+  function handleDelete(id: string) {
+    deleteDocument(id).then(() => {
+      updateDocs();
+    });
+  }
+
+
+  console.log(docs);
```
Comment (hunterhector) [inline src/app/pages/Project.tsx:41]: Remove console calls.

Comment (stevezhang1999) [reply src/app/pages/Project.tsx]: removed

File: src/app/pages/Ontology.tsx (hunk 6 above)
Comment (stevezhang1999) [reply src/app/pages/Ontology.tsx]: removed

File: simple-backend/nlpviewer_backend/models.py (hunk 4 above)
Comment (stevezhang1999) [reply simple-backend/nlpviewer_backend/models.py]: removed

File: simple-backend/nlpviewer_backend/handlers/project.py (hunk 3 above)
Comment (stevezhang1999) [reply simple-backend/nlpviewer_backend/handlers/project.py]: fixed

File: simple-backend/nlpviewer_backend/handlers/document.py (hunk 2 above)
Comment (stevezhang1999) [reply simple-backend/nlpviewer_backend/handlers/document.py]: fixed

File: simple-backend/nlpviewer_backend/handlers/document.py (hunk 1 above)
Comment (stevezhang1999) [reply simple-backend/nlpviewer_backend/handlers/document.py]: fixed

**Comment by stevezhang1999** [comment]
> There are a few high-level suggestions:
> 
> 1. Please reduce the number of warnings by typescript linting, for example, the following:
>    Expected '!==' and instead saw '!='
> 2. Can you include an example project in example db?
> 3. Please remove the "All Document" Link, and let's replace that with a "All Projects" link.
> 4. Please remove the ".idea" folder, the ".python-version" file, "electron-ssr-0.2.6.deb", and add them to your ".gitignore" to avoid them be included in the future.
> 5. Remove "console.log" and "print" comments after your own debug, which may prevent others from understanding the log output.
> 6. The main path "/" should probably be the same as "/projects", not it is the all documents page.

1. all '!=' and '==' are replaced.
2. new db.sqlite3.example is the new example db. It havs two example, projects, each has one examlple documents. The content of the example projects and documents are identical to the documents of the previous version.
3. done
4. done
5. done
6. done

File: simple-backend/nlpviewer_backend/models.py [hunk 8]
```diff
@@ -1,11 +1,36 @@
 from django.db import models
 
+class Project(models.Model):
+    name = models.CharField(max_length=200)
+    ontology = models.TextField(default='')
+
+    # relationship: document -- defined in Document
```
Comment (hunterhector) [inline simple-backend/nlpviewer_backend/models.py:7]: Coding style: commented discarded code can be removed during PR.

File: simple-backend/nlpviewer_backend/urls.py [hunk 9]
```diff
@@ -15,11 +15,14 @@
 """
 from django.contrib import admin
 from django.urls import include, path
-from nlpviewer_backend.handlers import session, user, document, nlp
+
+from nlpviewer_backend.handlers import session, user, document, project
+#from nlpviewer_backend.handlers import session, user, document, project, nlp
```
Comment (hunterhector) [inline simple-backend/nlpviewer_backend/urls.py:20]: Coding style: commented discarded code can be removed during PR.

File: simple-backend/nlpviewer_backend/urls.py [hunk 10]
```diff
@@ -44,10 +48,19 @@
...
+    path('projects/<int:project_id>/delete', project.delete),
+
     path('documents/<int:document_id>/text/edit', document.edit_text),   
 
-    path('nlp/load/<model_name>', nlp.load_model),
-    path('nlp/<int:document_id>/<model_name>', nlp.run_pipeline),
+    path('ontology_from_doc/<int:document_id>', document.get_doc_ontology_pack),
+
+    # path('nlp/load/<model_name>', nlp.load_model),
+    # path('nlp/<int:document_id>/<model_name>', nlp.run_pipeline),
```
Comment (hunterhector) [inline simple-backend/nlpviewer_backend/urls.py:63]: Coding style: commented discarded code can be removed during PR.

File: simple-backend/test.py [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+                                project = a)
+        
+
+    def test_project_relationship(self):
+        #"""Animals that can speak are correctly identified"""
+        doc1 = Document.objects.get(name="doc1")
+        project1 = Project.objects.get(name="project1")
+        project1.documents.first()
+        self.assertEqual(doc1, project1.documents.first())
+        self.assertEqual(project1, doc1.project)
```
Comment (hunterhector) [inline simple-backend/test.py:20]: New line.

File: src/app/App.tsx [hunk 12]
```diff
@@ -30,12 +37,15 @@ function App() {
 
           <nav>
             <ul>
-              <li>
+              {/* <li>
                 <Link to="/">All documents</Link>
-              </li>
+              </li> */}
```
Comment (hunterhector) [inline src/app/App.tsx:42]: Coding style: commented discarded code can be removed during PR.

File: src/app/lib/api.ts (hunk 5 above)
Comment (hunterhector) [reply src/app/lib/api.ts:69]: If this require another PR to solve, let's create an issue for this afterwards.

File: simple-backend/nlpviewer_backend/urls.py (hunk 10 above)
Comment (stevezhang1999) [reply simple-backend/nlpviewer_backend/urls.py:63]: The problem is the backend cannot start without forte and pytorch, is this by design?

Comment (hunterhector) [reply simple-backend/nlpviewer_backend/urls.py:63]: The current master has solved this issue. Now there will only be warnings if Forte is not installed.

File: src/app/App.tsx (hunk 12 above)
Comment (stevezhang1999) [reply src/app/App.tsx]: removed

File: simple-backend/test.py (hunk 11 above)
Comment (stevezhang1999) [reply simple-backend/test.py]: New line added. As well as some changes to conform the new Project and Document.

File: simple-backend/nlpviewer_backend/models.py (hunk 8 above)
Comment (stevezhang1999) [reply simple-backend/nlpviewer_backend/models.py]: Comments are rewritten.

File: simple-backend/nlpviewer_backend/handlers/project.py [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+    docs = project.documents.all().values()
+
+    return JsonResponse(list(docs), safe=False)
+
+@require_login
+def delete(request, project_id):
+    project = Project.objects.get(pk=project_id)
+    project.delete()
+
+    return HttpResponse('ok')
```
[the remark this answers is not in the document]
Comment (stevezhang1999) [reply simple-backend/nlpviewer_backend/handlers/project.py]: New line added

File: simple-backend/nlpviewer_backend/urls.py (hunk 10 above)
Comment (stevezhang1999) [reply simple-backend/nlpviewer_backend/urls.py:63]: So now should I delete them for the PR? or uncomment them? And the same problem for the above, 'import'

**Comment by stevezhang1999** [comment]
> 1. There are a few more linting problems (xxx is defined but never used), try to reduce them. We are likely to include linting in travis in the future. So let's try to reduce these now.
> 2. Another coding style problem, always adding new lines at the end of the file.
> 
> There are a few coding style suggestions in line.

1. fixed some "xxx is defined but never used", I will fix the remaining after finishing corresponding modules.
2. new line added and some commented code deleted.

File: simple-backend/nlpviewer_backend/urls.py (hunk 10 above)
Comment (hunterhector) [reply simple-backend/nlpviewer_backend/urls.py:63]: You can uncomment them now.

**Comment by hunterhector** [comment]
Looks like what's remaining are:
1. Some files need to be removed from the PR (the ones you have added to the .gitignore), but they should be removed from the PR as well.
2. Resolving some conflicts with the master.

I will see again if the PR is ready to merge afterwards.

**Comment by stevezhang1999** [comment]
> Looks like what's remaining are:
> 
> 1. Some files need to be removed from the PR (the ones you have added to the .gitignore), but they should be removed from the PR as well.
> 2. Resolving some conflicts with the master.
> 
> I will see again if the PR is ready to merge afterwards.


1.  I think all unnecessary  files are removed already.
2. I don't have the access to resolve conflicts after creating the PR, and the conflicts are
- db.sqlite3.example: The new example database
- DialogueBox.tsx: Only one change made for reducing linting warnings. More information in my comment.

**Comment by hunterhector** [comment]
> > Looks like what's remaining are:
> > 
> > 1. Some files need to be removed from the PR (the ones you have added to the .gitignore), but they should be removed from the PR as well.
> > 2. Resolving some conflicts with the master.
> > 
> > I will see again if the PR is ready to merge afterwards.
> 
> 1. I think all unnecessary  files are removed already.
> 2. I don't have the access to resolve conflicts after creating the PR, and the conflicts are
> 
> * db.sqlite3.example: The new example database
> * DialogueBox.tsx: Only one change made for reducing linting warnings. More information in my comment.

I just checked and the files are moved.  In terms of how to resolve the conflicts. I think you simply need to merge stave/master to your branch, you will be asked to resolve the merge conflicts, after that, simply push them.

**Comment by hunterhector** [comment]
Please make sure lining problems are resolved, right now the Travis check has failed due to unused variables.

**Comment by stevezhang1999** [comment]
> Please make sure lining problems are resolved, right now the Travis check has failed due to unused variables.

Now linting problems are resolved by deleting unfinished work.

**Comment by hunterhector** [comment]
Thanks for the hard work!

**Pull request merged**